Match realtime add

StephenHausler opened the pull request:
Adding new function to run Patch-NetVLAD in real-time from a web camera input. Also fixing a bug in match_two.py, where the cuda logic (line 196) was inverted meaning it was always defaulting to CPU rather than GPU.

Please review and merge @Tobias-Fischer @oravus :)

Tobias-Fischer commented on match_realtime.py:
```diff
@@ -0,0 +1,233 @@
+from patchnetvlad.tools import PATCHNETVLAD_ROOT_DIR
+
+
+def input_transform(resize=(480, 640)):
```
Why not `from patchnetvlad.tools.datasets import input_transform`?

StephenHausler replied:
Sounds good, ill do that

Tobias-Fischer commented on match_realtime.py:
```diff
@@ -0,0 +1,233 @@
+        ])
+
+
+def normalise_func(input_diff, num_patches, patch_weights):
```
Same here, import the method.

StephenHausler replied:
Ok great, thanks ill do that

Tobias-Fischer commented on match_realtime.py:
```diff
@@ -0,0 +1,233 @@
+    return normed_diff
+
+
+def plot_two(cv_im_one, cv_im_two, inlier_keypoints_one, inlier_keypoints_two):
```
And same here - generalize the existing plot_two so it can be used in `match_two.py` and here.

StephenHausler replied:
Ok sounds good

Tobias-Fischer commented on match_realtime.py:
```diff
@@ -0,0 +1,233 @@
+    cv2.imshow('frame', im_allpatch_matches)
+
+
+def match_two(model, device, opt, config, im_one, im_two):
```
And again, import from the existing code, and make more general if needed.

StephenHausler replied:
This should be possible to do as well, I'll make a new commit with these changes to the pull request

Tobias-Fischer reviewed (changes requested):
Hi @StephenHausler - many thanks for that! A few times we can just re-use existing code rather than duplicating it, it will make maintenance easier.

StephenHausler commented:
Made review changes @Tobias-Fischer. Could not easily use patchnetvlad.tools.datasets import input_transform due to being a class method, but still made neater by importing from match_two. match_realtime only has a main function now, all other required functions are imported.

Tobias-Fischer commented:
Ah. Let's make it work then and fix `match_two` as well. Simply move `input_transform` out of the class and make it a normal function that can be imported. Similarly, let's pull in normalise_func and not duplicate code. Finally, can you amend https://github.com/QVPR/Patch-NetVLAD/blob/main/setup.py#L57 to include the new script please :)

StephenHausler commented:
Ok great thanks, fixed up input_transform and setup.py. I renamed the normalization function in match_two to apply_patch_weights, the two functions (the norm function in local_matcher vs the one in match_two) were doing different things and the old name was misleading.

Tobias-Fischer commented:
Looking good to me. @oravus - do you mind having a look and merge?

oravus reviewed (approved):
Have read through, looking good!
